### ISSUE TITLE
take into account the churn rate

### ISSUE THREAD

**Issue by dachary** [EliotBerriot/mnm#16]
The https://en.wikipedia.org/wiki/Churn_rate tells us how many instances / users are active. It differentiates between a user creating an account and not doing anything and a user actively participating. It tells us how long user stay on mastodon after the initial buzz.

1. Why are you opening the issue for: the churn rate / attrition rate needs to be taken into account for the stats to be more meaningful. 

Thanks !

**Comment by EliotBerriot** [comment]
Measuring the engagement of users is indeed important though I'm not sure how to obtain this data in a reliable and scalable manner.

The best way would be to have this data available directly available on each instance API, but to my knowledge, this is not the case at the moment. Also, we'd have to define what is exactly an active account. An account that sends at least one toot in the previous 30 days?

I think this may worth a pull request on Mastodon repository, because this stats can also be interesting for instances admin and community managers.

**Comment by dachary** [comment]
It's indeed a difficult problem and it requires adjustments on a regular basis. It is however easy to do simple things such as what you suggest. An account that never did anything can be discarded entirely and immediately. An account that had no activity at all in the past 30 days could be considered inactive. An account that boosts / favorits / follows but never toots should probably be considered differently from an account that toots (i.e. active vs passive).

Although showing an ever growing number of accounts is good and useful right now, a few weeks after the expanse, it will quickly become meaningless as the users come and go. Our strength is going to be measured in the number of active user at a given time. I would very much like to know how large is our community :-)

**Comment by EliotBerriot** [comment]
I'll open an issue on the main Mastodon repository to explain the use case and see what can be done.

**Comment by gled-rs** [comment]
Counting the local active users in the past 24hours:
SELECT COUNT(distinct a.username) FROM statuses AS s, accounts AS a WHERE a.id=s.account_id AND a.domain IS NULL and s.created_at >= now() - interval '24 hour'

You can adapt that to check the MAU, WAU etc... ;)

**Comment by dachary** [comment]
requiring changes in mastodon to obtain more information does not seem to be a practical solution. Hundreds of instances out there don't have them and waiting long enough for all instances to upgrade won't work very well ;-) statistics should be based on publicly available information such as the page of existing users listing followers, activity etc.

**Comment by EliotBerriot** [comment]
@gled-rs Yeah, this is possible, but in order to run this query, you have to actually have access to each Mastodon instance's database which is not the case. I can only rely on public API or data displayed in public HTML pages, such as https://mastodon.xyz/about/more

@dachary unless I'm missing something, there is no API endpoint offering a full listing of user activity (that but be too huge to handle for servers, and could cause privacy / spamming issues).

I agree with you when you say that implementing this in MAstodon public API will take some time bcause of instance upgrade. Improving Mastodon itself is not a short-term solution, however on the long term, it's the most scalable approach, because we don't have to "guess" the data, it's already offered by each instance. This also means other projects could benefit from this.

**Comment by dachary** [comment]
@EliotBerriot the only way to get stats is via the API indeed. A bot should read user pages, remember where it left of, read the list of followers, the status, read more user pages. And maintain a stats database from the collected information. And from that database queries to figure out how many active users / instances we have is likely to be fairly accurate.

P.S. I'm not proposing a bot following users, that would be the wrong way of doing it ;-)

**Comment by EliotBerriot** [comment]
> A bot should read user pages, remember where it left of, read the list of followers, the status, read more user pages. And maintain a stats database from the collected information. And from that database queries to figure out how many active users / instances we have are likely to be fairly accurate.

This would work, but this cannot really scale. In a very optimistic scenario, where you do 1 HTTP request per user, if you wanted to parse all Mastodon users account every day, this would require doing more than 500k HTTP requests a day to get this data. Since you're talking about parsing also followers and followees, this would mean even more requests per-user, so let's say 3 HTTP requests per user (I think this is really, really optimistic). So 1.5 million HTTP requests a day, given the current size of the network (which will only grow).

Now, if you need 2 seconds to make a single HTTP request, this means a single worker could process ~40K users a day. So you'd have to spawn 10s of workers to do this job, which would require many servers. 

And the load on carwled instances would be really big, too, leading in blacklisting / throttling.

**Comment by dachary** [comment]
There certainly is a trade-off between absolute accuracy and very obsolete or no information at all. Right now a 50% sampling over a period of days (24h is probably overkill) is more than enough to get accurate stats. Unless the network grows in a way that is 100% random, reducing the sample % in proportion should work just fine. In other words if N% of the users have been sampled over the past few days, it does not matter much that 100-N% have not been sampled recently.

**Comment by EliotBerriot** [comment]
There are other issues caused by this approach. Crawling individual profiles mean we store profile information, at least the username and the activity. This means maintaining a huge database of users along with their statistics. Inevitably, some users will want to opt-out from this, meaning we will also need to manage that, adding complexity to the whole system.

If we do this with aggregated data, at the *instance* level, their are no such concerns. This is more efficient by many orders of magnitude, a single worker could probably handle the whole network right now, it does not raise any privacy issue, and it benefits to the whole network by making this data available to anybody (and not just MNM).

**Comment by dachary** [comment]
In that case I suppose we can close this as "Won't fix" and delegate resolution of this problem to Mastodon.

**Comment by EliotBerriot** [comment]
**I'm sorry if it was too abrupt, that was not my intention, and since english is not my mother tongue, I may end up being less nuanced than I actually am.**

We can leave this open, because the initial concern ("measuring the churn rate") is absolutely valid and has its place within the project. I was just disagreeing with you regarding the implementation, as I don't think it's realistic to develop, maintain and deploy such a crawler. 

If you are okay with that, I will open the issue on Mastodon, requesting the following data to be available in the public API:

- Publishing users: users who published a toot in the last N days
- Sharing users: users who boosted or favorited a toot in the last N days
- Active users: users with at least one of the previously mentioned activities
- Inactive users: users without any of the previously mentioned activities

We will need to define the ideal amount for N, maybe we could have many (7 days, 30 days and 60 days, for example). Once merged in Mastodon, this will probably take a few weeks before a mojority of instances actually do the upgrade, but I believe it can be pretty fast: the network is still young, and because of its infancy bugs, admins are deploying updates faster than they usually do.

Once deployed at the instance level, it would be really easy to gather and display this data in MNM.

Do you think this could solve your issue, even if it takes longer than directly crawling user profiles? If yes, please reopen the issue, I'll open the related issue on the Mastodon repository.

**Comment by dachary** [comment]
I was not offended, thanks very much for your apology, this is very gracious of you :-) The path you propose makes sense.

**Comment by EliotBerriot** [comment]
A [related issue](https://github.com/tootsuite/mastodon/issues/874) so I just added an answer including important points.
